src/features/click-to-play.js

```diff
@@ -1411,17 +1403,32 @@ export function init (args) {
     })
 
     // Request the current state of Click to Load from the platform.
-    // Note: When the response is received, initCTL() is then called by the
-    //       response handler to finish starting up the feature.
+    // Note: When the response is received, the response handler finishes
+    //       starting up the feature.
     sendMessage('getClickToLoadState')
 }
 
-export function update (args) {
-    const detail = args && args.detail
-    if (!(detail && detail.func)) { return }
+export function update (message) {
+    // TODO: Once all Click to Load messages include the feature property, drop
+    //       messages that don't include the feature property too.
+    if (message?.feature && message?.feature !== 'clickToLoad') return
 
-    const fn = updateHandlers[detail.func]
-    if (typeof fn !== 'function') { return }
+    const messageType = message?.messageType
+    if (!messageType) return
 
-    fn(detail.response)
+    // Message responses.
+    if (messageType === 'response') {
+        const messageResponseType = message?.responseMessageType
+        if (messageResponseType && knownMessageResponseType(messageResponseType)) {
+            return messageResponseHandlers[messageResponseType](message.response)
+        }
+    }
+
+    // Other known update messages.
+    if (messageType === 'displayClickToLoadPlaceholders') {
```

Comment:
The issue with this approach is that it will cause us to re-enumerate the DOM for each tracker that we block at https://github.com/duckduckgo/duckduckgo-privacy-extension/pull/1738/files#diff-ed5e7ba3132fdd67f6264c179d3641e8959578d350a29280c36f58ab228c52f5R260, which can result in a lot of redundant enumeration.  For https://privacy-test-pages.glitch.me/privacy-protections/click-to-load/ we do 8 passes when 1 would do.

A mitigation might be to do a set a semaphore that a refresh is needed (f not already set), and then do a setTimeout for like 10ms, fire `replaceClickToLoadElements()` and clear the semaphore.  That way we'd enumerate likely once (or every 10ms at most) vs repeatedly for a number of blocked elements at load.

Reply:
Yea, the test pages are pretty much the worst case for this approach. A bunch of blocked elements, on one page, that were all ready before page load, so the event fires multiple times but there are not more elements found each time. Usually though, this new behaviour is what we want. We don't want to wait for page load to display placeholders, nor assume that because we didn't find any blocked elements in the past, that there aren't any in the future.

I agree with your suggested mitigation, if performance proves to be an issue I'd be up for doing something like that to reduce the number of refreshes that happen in quick succession. I'd prefer not to add that for now though, I figure there's a good chance it's a case of premature optimisation. (I would also take more care to use the provided ruleAction at that point.) What do you think?

Reply:
For FB perhaps but for YT pages of videos aren't that uncommon.  Take https://www.surfingpaddling.com/dropping-in-surf-show for example.  Granted its already a slow page to load but with YT protections enabled it basically ceases to render (for me, in Chrome).

Reply:
_(As discussed in MatterMost, it turned out the (broken) `hideTrackingElement` option was causing an exponential number of placeholder elements to be added. Which caused websites like that with a large number of videos to be slow or crash. I have disabled that option for now, since I have a follow-up task to remove/fix it already anyway.)_

Reply:
I spent some time profile this with and without a setTimeout hack, and didn't notice much difference so I think we're good to go here!